### DISCUSSION TITLE
checking for different versions of python 🐍🐍🐍

### DISCUSSION THREAD

**Discussion by ipatch** [Homebrew/discussions#2750] [Tap maintenance and Homebrew development]
is there a good way to check for different versions of python? ie. specifically if a `$USER` has both `python@3.9` and `python@3.10` installed?  i'm aware that `python@3.10` is keg-only for the time being. but i've run into some [use cases ](https://github.com/FreeCAD/homebrew-freecad/issues/267)where people have installed `python@3.10` and then attempt to install freecad provided by the homebrew-freecad tap, and unfortunately `pyside2` will not compile/work with `python@3.10`.

1. is it possible for formula authors to see which homebrew python is currently being used. ie. the homebrew python that is currently linked. so say `$USER_a` has somehow run `brew link python@3.10` and is attempting to install freecad, and knowing that freecad will not build using `python@3.10` a graceful error is presented saying something like _not compatible_ with `python@3.10` please _link_ or _relink_ or `python@3.9`.

the closest thing i came across reading the formula cookbook on the docs.brew.sh site was using the `conflicts_with` block, but it seemed to be last resort type of use, so not exactly sure if this is what i need to use. 🧐

**Comment by gromgit** [comment]
> people have installed `python@3.10` and then attempt to install freecad provided by the homebrew-freecad tap, and unfortunately `pyside2` will not compile/work with `python@3.10`

I see from https://github.com/FreeCAD/homebrew-freecad/blob/master/Formula/pyside2.rb that you're already telling CMake the fully-qualified Python binary to use. If that's not working, and `pyside2` is still being built with whatever Python is actually linked, you can try prepending the correct Python's path in the formula instead:
```
ENV.prepend_path "PATH", "#{pyhome}/bin"
```
Similarly, if any formula absolutely requires `python@3.9` at runtime, I believe the canonical solution is to use `write_env_script` to create a script wrapper that does the necessary `PATH` prepending. See the `autojump` formula for a worked example.

**Reply by ipatch** [reply]
thanks, i'll take a look at the formula sometime tomorrow. and if i can get a working example going, i'll mark my question as solved.

thank you. 🙏

also the `freecad.rb` formula should be using the `pyside2@5.15.2.rb` formula within the repo, and **not** the `pyside2.rb` that's a formula that i need to phase out.
